### PR TITLE
hooks may select a user to run as

### PR DESCRIPTION
**Pull request by anthonybishopric**

### CODE REVIEW

File: pkg/pods/pod_manifest.go [hunk 1]
```diff
@@ -28,7 +28,7 @@ type LaunchableStanza struct {
 type PodManifest struct {
 	Id                string                      `yaml:"id"` // public for yaml marshaling access. Use ID() instead.
 	LaunchableStanzas map[string]LaunchableStanza `yaml:"launchables"`
-	Config            map[string]interface{}      `yaml:"config"`
+	Config            map[interface{}]interface{} `yaml:"config"`
```
Comment (tummychow) [inline pkg/pods/pod_manifest.go]: was this change actually necessary to get the rest of this to work?

Comment (anthonybishopric) [inline pkg/pods/pod_manifest.go]: yes, because of [this line](#diff-bfa12c9e8482818801371a2b62106430R162)

Comment (tummychow) [inline pkg/pods/pod_manifest.go]: Yeah but why can't LoadFromUnpacked just take a `map[string]interface{}` instead?

Comment (anthonybishopric) [inline pkg/pods/pod_manifest.go]: There's nothing wrong with an application specifying an integer key in their own config, so I'd rather not coerce it to a string.

Comment (tummychow) [inline pkg/pods/pod_manifest.go]: In general I don't like using interfaces as keys; if they're uncomparable you get a panic at time of indexing. It doesn't really help that YAML is a weird format and might allow unusual key types (I read the YAML spec front to back once and still don't understand it). The impact is small here and the risk is unclear, so I won't block the PR on this, but personally I would rather enforce string keys everywhere.

**Comment by tummychow** [comment]
looks like some typos in the commit message, "may select"? otherwise :+1:

**Comment by petertseng** [comment]
I think you may want to remove https://github.com/square/p2/blob/master/pkg/preparer/listener.go#L88 (this line, far as I can tell, comes after you set `runAs` at https://github.com/square/p2/pull/118/files#diff-bfa12c9e8482818801371a2b62106430R199)

File: pkg/pods/pod_manifest.go (hunk 1 above)
Comment (anthonybishopric) [inline pkg/pods/pod_manifest.go]: YAML is confusing enough, us adding a new constraint on the otherwise freeform YAML `config` stanza is going to be deeply unintuitive. Agreed that we should discourage it though, it *is* weird.

**Comment by anthonybishopric** [comment]
Wow that commit message is really messed up huh

**Comment by anthonybishopric** [comment]
@petertseng good catch, will change.

File: pkg/preparer/listener.go [hunk 2]
```diff
@@ -79,13 +79,23 @@ func (l *HookListener) Sync(quit <-chan struct{}, errCh chan<- error) {
...
-			curUser, err := user.Current()
+
+			// Hooks may specify a user to run as - by default they will run as their pod ID.
+			topLevelConfig := config.LoadFromUnpacked(result.Manifest.Config)
+			hookConfig, err := topLevelConfig.ReadMap("hook")
+			if err != nil {
+				sub.WithField("err", err).Errorln("Hook configuration is invalid")
+				break
+			}
+			runAs, err := hookConfig.ReadString("run_as")
```
Comment (stouset) [inline pkg/preparer/listener.go]: Seems like it'd be useful to be able to specify nested paths in one call.

Comment (anthonybishopric) [inline pkg/preparer/listener.go]: We had talked about a dot syntax for that at one point

`config.ReadString("foo.bar.baz")`

May come back to that in the future
